[terra-form-validation] Update form validation docs, add FormValidationUtil, mark package as public. (#635)
cerner/terra-framework: packages/terra-form-validation/docs/NumericValidation.md

              inputId="numExample"
              label="Numeric Example"
              error={meta.error}
              isInvalid={meta.error !== undefined}
              inputAttrs={{
                type: 'number',
                ...input,
              }}
              onChange={(e) => { input.onChange(e.target.value); }}
              value={input.value}
              required
            />
          )}
        </Field>
        <Button text="Submit" type={Button.Opts.Types.SUBMIT} />
      </form>
    );
  }

  render() {
    return (
      <Spacer marginBottom="small">
        <Form
          onSubmit={this.submitForm}
          render={this.renderForm}
          initialValues={{ numExample: '' }}
          validate={(values) => {
            const errors = {};

            if (!values.numExample) {
              errors.numExample = 'Required';
            }

            return errors;
          }}
        />
        {this.state.submittedValues
          && (
          <div>
            <p>Form Submitted Successfully With</p>
            <pre>{JSON.stringify(this.state.submittedValues, 0, 2)}</pre>
          </div>
          )
        }
      </Spacer>
    );
  }
}
```

> With this, we have a pretty simple number type input with a basic validation tied to the `<Form>` component that checks to see whether or not we have a value for the input.

Reviewer: ```suggestion
We have a simple number type input with validation, tied to the `<Form>` component, that checks whether we have a value for the input.
```

We also have our `<Field>`'s validation function defined and ready to be added to, with it only having a basic check to see if there is a value.

> Because we have added to the `inputAttrs` prop  of `<InputField>` the value of `type: 'number'` this also means we have a bit of browser enforcement on the inputted value. Mainly that it has to be a number of some sort so it only accepts certain characters.

Reviewer: ```suggestion
We have added browser enforcement on the inputted value by passing `type: 'number'` into `<Inputfield>`'s `inputAttrs`.
```

> A thing of note is that even though we specified the input as a number type, it will still take some 'wrong' values and place them in the field but give them as `undefined`. Specifically the input will accept values such as `--1` or `2.0.32` but then when the validation functions run the value given to them will be `undefined`. With our code is currently setup that means we will just return a value of `Required` via the `<Form>`'s validation function.

Reviewer: ```suggestion
Note that specifying `type: 'number'` still allows negatives and decimal points, such as `--1` or `2.0.32`. These will be placed in the field as `undefined`. With our code, we will just return a value of `Required`, via the `<Form>`'s validation function, if values are undefined.
```

> The 2nd thing of note is that our validation function, `validateNumber`, will be given it's argument in the form of a string despite it being a number type input. This doesn't impact simple operators too much like `<` and `>` but it is information you should know for when your building custom validation functions.

Reviewer: ```suggestion
Second thing - our validation function, `validateNumber`, will be given a *string*, despite the input being of `type: 'number'`. This does not impact simple operators but is useful to know when building custom validation functions.
```

> With this in mind, we'll first start by giving a minimum and maximum to our value. For simplicity's sake lets go with a minimum of 10 and a maximum of 100.

Reviewer: ```suggestion
Let's start by giving a minimum and maximum to our value. For simplicity, set a minimum to 10 and a maximum to 100.
```

```javascript
const validateNumber = (value) => {
  if (!value) { return undefined; }
  if (value < 10) {
    return 'Value should not be less than 10.';
  }
  if (value > 100) {
    return 'Value should not be over 100.';
  }

  return undefined;
};
```

> Now for the more complicated validation which is the level of precision we want in our input. For the example let us say that we want 3 decimals worth of precision and we don't want to accept more than that. There are multiple ways of validating this but since our value is a string, we'll use string functions to validate.

Reviewer: ```suggestion
Now, let's check for precision. For this example, lets validate to the thousandth place (three decimal places). Since our value is a string, we'll use string functions to validate.
```

> We will split the given string on its `.` character, and then check to make sure the 2nd string we get from doing this has a length of 3 or less.

Reviewer: ```suggestion
Use `.` as a delimiter and check for a length less than or equal to 3.
```

```javascript
const validateNumber = (value) => {
  ...
  const valueSplit = value.split('.');
  if (valueSplit.length === 2 && valueSplit[1].length > 3) {
    return 'Value has more than 3 decimal points';
  }

  return undefined;
};
```

> This way of validating for precision has a few quirks with it, mainly that if we give it a value of `23.023000` it would fail the validation, because of the extra `0`s at the end despite this number effectively being the same as `23.023`.

Reviewer: This validation has a quirk. If the input contains extra `0`s, like `23.0230000`, validation will fail.

Author: Swapped https://github.com/cerner/terra-framework/pull/635/commits/a0241f8008eacb2cb9bba04c762e6018716f0ff5

> For the sake of the example lets try a different way of validating. Let us, instead of using string function to validate, convert the value given to us to a `Number` instead. With it being a number we can use the `Number` class function `toFixed` which allows us to automatically take a value to a given precision.

Reviewer: ```suggestion
Lets try a different way of validating. Convert the value a `Number` instead. This way we can use the `toFixed` function to automatically set precision, and use this as a comparison.
```

```javascript
const validateNumber = (value) => {
  ...
  const valueNum = Number(value);
  if (valueNum.toFixed(3) != valueNum) {
    return 'Value has more than 3 decimal points';
  }

  return undefined;
};
```

> With this we compare the value given us to same value to a given precision. Which means if we are handed `45.4321`, we will compare it to `45.432`, which is not equal and will return the failure message. The quirk with this validation is that if you give it `23.023000` it will succeed because it will be compared to `23.023` and be found equal thus you probably will need to include something to shorten the String value in this case.

Reviewer: This validation has the opposite quirk. If the input contains extra 0s, like 23.0230000, validation will succeed.

Author: Swapped https://github.com/cerner/terra-framework/pull/635/commits/a0241f8008eacb2cb9bba04c762e6018716f0ff5

          </div>
          )
        }
      </Spacer>
    );
  }
}
```

We have a simple number type input with validation, tied to the `<Form>` component, that checks whether we have a value for the input.

We also have our `<Field>`'s validation function defined and ready to be added to, with it only having a basic check to see if there is a value.

We have added browser enforcement on the inputted value by passing `type: 'number'` into `<Inputfield>`'s `inputAttrs`.

Note that specifying `type: 'number'` still allows negatives and decimal points, such as `--1` or `2.0.32`. These will be placed in the field as `undefined`. With our code, we will just return a value of `Required`, via the `<Form>`'s validation function, if values are undefined.

Second thing - our validation function, `validateNumber`, will be given a *string*, despite the input being of `type: 'number'`. This does not impact simple operators but is useful to know when building custom validation functions.

Let's start by giving a minimum and maximum to our value. For simplicity, set a minimum to 10 and a maximum to 100.

```diff
const validateNumber = (value) => {
  if (!value) { return undefined; }

> + if (value < 10) {

Reviewer 2: nit: I feel like we should be casting `value` to a number data type here: `if (Number(value) < 10)` even though this will work without it.

Author: Fixed here https://github.com/cerner/terra-framework/pull/635/commits/415c46ab9a0b74893a686c38430b1d8decd708ad#diff-efb0b388a04429df36f1ce4eb79ef872R108
The other references to this code later in the example should also be fixed.

+   return 'Value should not be less than 10.';
+ }
+ if (value > 100) {
+   return 'Value should not be over 100.';
+ }

  return undefined;
};
```

Now, let's check for precision. For this example, lets validate to the thousandth place (three decimal places). Since our value is a string, we'll use string functions to validate.

Use `.` as a delimiter and check for a length less than or equal to 3.

```diff
const validateNumber = (value) => {
  ...
+ const valueSplit = value.split('.');
+ if (valueSplit.length === 2 && valueSplit[1].length > 3) {
+   return 'Value has more than 3 decimal points';
+ }

  return undefined;
};
```

This validation has a quirk. If the input contains extra `0`s, like `23.0230000`, validation will fail.

> Lets try a different way of validating. Convert the value a `Number` instead. This way we can use the `toFixed` function to automatically set precision, and use this as a comparison.

Reviewer 2: 👍

  ...
- const valueSplit = value.split('.');
- if (valueSplit.length === 2 && valueSplit[1].length > 3) {
+ const valueNum = Number(value);
+ if (valueNum.toFixed(3) != valueNum) {
    return 'Value has more than 3 decimal points';
  }

  return undefined;
};
```

This validation has the opposite quirk. If the input contains extra `0`s, like `23.0230000`, validation will succeed.

With this we have a function that we can add to validate a number based by having a minimum, maximum, and a certain level of precision required. In the same vein as the 'Max Length Input' example, we can use the `FormValidationUtil` to help control and reduce some of workflow.

```diff
const validateNumber = (value) => {
-  if (value < 10) {
+  if (!FormValidationUtil.isOverMinValue(value, 10)) {
    return 'Value should not be less than 10.';
  }
-  if (value > 100) {
+  if (!FormValidationUtil.isUnderMaxValue(value, 100)) {
    return 'Value should not be over 100.';
  }
- const valueNum = Number(value);
- if (valueNum.toFixed(3) != valueNum) {
+ if (!FormValidationUtil.isPrecise(value, 3)) {
    return 'Value has more than 3 decimal points';
  }

  return undefined;
};
```

> Note: `FormValidationUtil.precisionCheck` uses the first method check precision.

Reviewer 2: Hmm, Im not sure what this means.

Author: Cleared it up here https://github.com/cerner/terra-framework/pull/635/commits/415c46ab9a0b74893a686c38430b1d8decd708ad#diff-efb0b388a04429df36f1ce4eb79ef872R108
